api/graphite.go

```diff
@@ -372,6 +373,84 @@ func (s *Server) metricsFind(ctx *middleware.Context, request models.GraphiteFin
 	}
 }
 
+func (s *Server) metricsExpand(ctx *middleware.Context, request models.GraphiteExpand) {
+	var wg sync.WaitGroup
+	var errAtomic atomic.Value
```

Comment:
we dont need to use atomic values here.  just use `golang.org/x/sync/errgroup`
eg
```
g, reqCtx := errgroup.WithContext(ctx.Req.Context())
results := make([]map[string]struct{}, len(request.Query))
for i, query := range request.Query {
	i, query := i, query
	g.Go(func() error {
		series, err := s.findSeries(reqCtx, ctx.OrgId, []string{query}, 0)
		if err != nil {
			return err
		}
		results[i] = make(map[string]struct{})
		for _, s := range series {
			for _, n := range s.Series {
				if request.LeavesOnly && !n.Leaf {
					continue
				}

				results[i][n.Path] = struct{}{}
			}
		}
		return nil
	})
}
if err := g.Wait(); err != nil {
	response.Write(ctx, response.WrapError(err))
	return
}

// check to see if the request has been canceled, if so abort now.
select {
case <-reqCtx.Done():
	//request canceled
	response.Write(ctx, response.RequestCanceledErr)
	return
default:
}
....
```

Reply:
thx, forgot about errgroup. i used it now: https://github.com/grafana/metrictank/pull/1795/commits/5b596a6085f0d58623639455873cc8778692064f
i had to separate the errgroup's ctx from the one that gets checked after the errgroup is done, because when it's done it cancels the ctx.